src/components/ripple/ripple-manager.ts

```diff
@@ -0,0 +1,174 @@
+import {
+  ElementRef,
+} from '@angular/core';
+
+/** @internal */
+export enum ForegroundRippleState {
+  NEW,
+  EXPANDING,
+  FADING_OUT,
+}
+
+/**
+ * Wrapper for a foreground ripple DOM element and its animation state.
+ * @internal
+ */
+export class ForegroundRipple {
+  state = ForegroundRippleState.NEW;
+  constructor(public rippleElement: Element) {}
+}
+
+const RIPPLE_SPEED_PX_PER_SECOND = 1000;
+const MIN_RIPPLE_FILL_TIME_SECONDS = 0.1;
+const MAX_RIPPLE_FILL_TIME_SECONDS = 0.3;
+
+const sqr = (x: number) => x * x;
+
+/**
+ * Returns the distance from the point (x, y) to the furthest corner of a rectangle.
+ */
+const distanceToFurthestCorner = (x: number, y: number, rect: ClientRect) => {
+  const maxSquaredDistance = Math.max(
+      sqr(x - rect.left) + sqr(y - rect.top),
+      sqr(rect.right - x) + sqr(y - rect.top),
+      sqr(x - rect.left) + sqr(rect.bottom - y),
+      sqr(rect.right - x) + sqr(rect.bottom - y));
+  return Math.sqrt(maxSquaredDistance);
```

Comment:
That is better, and we can get rid of the custom sqr function using Math.abs. Also O(N) instead of O(2^N), which will be nice if we ever want to render 26-dimensional ripples to support superstrings.